src/vmspawn/vmspawn-util.c

```diff
@@ -236,3 +251,35 @@ int find_qemu_binary(char **ret_qemu_binary) {
 
         return find_executable(qemu_arch_specific, ret_qemu_binary);
 }
+
+int machine_cid(unsigned *ret_machine_cid) {
+        int r;
+        uint64_t cid;
+        _cleanup_close_ int vfd = -EBADF;
+
+        /* Enumerate all machine CIDs in [3..UINT_MAX) until we find a free CID.
+         * if there are no free addresses return SYNTHETIC_ERRNO(EADDRNOTAVAIL)
+         * if ioctl returns an unexpected error forward that
+         */
+
+        r = open("/dev/vhost-vsock", O_RDWR | O_CLOEXEC);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open /dev/vhost-vsock as read/write: %m");
+        vfd = r;
+
+        for (cid = 3; cid < UINT_MAX; cid++) {
```

Comment:
hmm, I think the start value should probably be hashed from the machine name (i.e. `--machine=`, like nspawn has it), so that it is automatically stable. The 32bit namspace is a bit too small to make this fully reliable, but it's better than nothing. And if the value is already takne, maybe pick another random value?

and leave the bottom 1K or so free, for people to assign manually?

Reply:
Vmspawn currently doesn't have a --machine argument, will the image name do for now?
mkosi does this by hashing the entire "config" object so basically all of the command line parameters.
The current implementation concept is from libvirt.

Reply:
just add --machine=. it's trivial.

i mean, sooner or later we want this anyway (so that we can register the vm with machined), might as well add this now, even if it isn't used for anything yet but the hashing of the cid

Reply:
I have added --machine=. And implemented a fairly basic way of deterministically generating 1000 random CIDs using siphash so that the n-th vmspawn instance should always get the n-th random CID. This falls back to random after 1000 attempts of deterministic CIDs.

Reply:
Fixed in 7126b0eb12.